Cabal/tests/PackageTests/Tests.hs

```diff
@@ -0,0 +1,157 @@
+      r <- shouldFail $ cabal "configure" []
+      -- Should tell you how to enable the desired behavior
+      let sb = "library which is defined within the same package."
+      assertOutputContains sb r
+
+  , tc "BuildDeps/InternalLibrary1" $ cabal_build []
+  , tc "BuildDeps/InternalLibrary2" $ internal_lib_test "internal"
+  , tc "BuildDeps/InternalLibrary3" $ internal_lib_test "internal"
+  , tc "BuildDeps/InternalLibrary4" $ internal_lib_test "installed"
+  , tc "BuildDeps/SameDepsAllRound" $ cabal_build []
+  , tc "BuildDeps/TargetSpecificDeps1" $ do
+      cabal "configure" []
+      r <- shouldFail $ cabal "build" []
+      assertBool "error should be in MyLibrary.hs" $
+          resultOutput r =~ "^MyLibrary.hs:"
+      assertBool "error should be \"Could not find module `System.Time\"" $
+          resultOutput r =~ "Could not find module.*System.Time"
+  , tc "BuildDeps/TargetSpecificDeps2" $ cabal_build []
+  , tc "BuildDeps/TargetSpecificDeps3" $ do
+      cabal "configure" []
+      r <- shouldFail $ cabal "build" []
+      assertBool "error should be in lemon.hs" $
+          resultOutput r =~ "^lemon.hs:"
+      assertBool "error should be \"Could not find module `System.Time\"" $
+          resultOutput r =~ "Could not find module.*System.Time"
+  , tc "PathsModule/Executable" $ cabal_build []
+  , tc "PathsModule/Library" $ cabal_build []
+  , tc "PreProcess" $ cabal_build ["--enable-tests", "--enable-benchmarks"]
+  , tc "PreProcessExtraSources" $ cabal_build ["--enable-tests", "--enable-benchmarks"]
+  , tc "TemplateHaskell/vanilla" $ cabal_build []
+  , tc "TemplateHaskell/profiling" $ cabal_build ["--enable-library-profiling", "--enable-profiling"]
+  , tc "TemplateHaskell/dynamic" $ cabal_build ["--enable-shared", "--enable-executable-dynamic"]
+  , tc "CMain" $ cabal_build []
+
+  -- See https://github.com/haskell/cabal/issues/1241
+  , tc "EmptyLib" $
+      withPackage "empty" $ cabal_build []
+
+  , tc "Haddock" $ do
+      dist_dir <- distDir
+      let haddocksDir = dist_dir </> "doc" </> "html" </> "Haddock"
+      cabal "configure" []
+      cabal "haddock" []
+      let docFiles
+              = map (haddocksDir </>)
+                    ["CPP.html", "Literate.html", "NoCPP.html", "Simple.html"]
+      -- TODO: refactor me
+      mapM_ (assertFindInFile "For hiding needles.") docFiles
+
+  , tc "OrderFlags" $ cabal_build []
+
+  , tc "ReexportedModules" $ do
+      whenGhcVersion (>= Version [7,9] []) $ cabal_build []
+
+  , tc "UniqueIPID" . withPackageDb $ do
+      withPackage "P1" $ cabal "configure" []
+      withPackage "P2" $ cabal "configure" []
+      withPackage "P1" $ cabal "build" []
+      withPackage "P1" $ cabal "build" [] -- rebuild should work
+      r1 <- withPackage "P1" $ cabal "register" ["--print-ipid", "--inplace"]
+      withPackage "P2" $ cabal "build" []
+      r2 <- withPackage "P2" $ cabal "register" ["--print-ipid", "--inplace"]
+      -- TODO: hella janky
```

Comment:
You may want to golf it, but it's *test code* and it's reasonably *understandable*, so remove the TODO. If someone wants to golf it or extract some common logic at some point, they will...